ScopeNotActiveException: Injection scope io.helidon.service.registry.Service.Singleton[7] is not active.

## Environment Details
* Helidon Version: 4.2.2
* Helidon SE
* JDK version: Corretto-24.0.1.9.1 (build 24.0.1+9-FR)
* OS: Debian Linux

----------

## Problem Description
We get the following error when execution tests where for each test case a custom `ServiceRegistryManager` gets started and shut down.
```
19:07:17.173 [[0x3bda7cb6 0x175d2c6e] WebServer socket] WARN io.helidon.webserver.http1.Http1Connection - Internal server error
io.helidon.http.RequestException: Internal error
	at io.helidon.http.RequestException$Builder.build(RequestException.java:139)
	at io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:249)
	at io.helidon.webserver.ConnectionHandler.run(ConnectionHandler.java:175)
	at io.helidon.common.task.InterruptableTask.call(InterruptableTask.java:47)
	at io.helidon.webserver.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:239)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:466)
Caused by: io.helidon.service.registry.ScopeNotActiveException: Injection scope io.helidon.service.registry.Service.Singleton[7] is not active.
	at io.helidon.service.registry.ScopedRegistryImpl.checkActive(ScopedRegistryImpl.java:164)
	at io.helidon.service.registry.ScopedRegistryImpl.activator(ScopedRegistryImpl.java:136)
	at io.helidon.service.registry.ServiceManager.activator(ServiceManager.java:81)
	at io.helidon.service.registry.ServiceSupplies.oneInstance(ServiceSupplies.java:42)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.lambda$new$1(ServiceSupplies.java:159)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.get(ServiceSupplies.java:170)
	at io.helidon.service.registry.CoreServiceRegistry.get(CoreServiceRegistry.java:180)
	at io.helidon.service.registry.CoreServiceRegistry.get(CoreServiceRegistry.java:175)
	at io.helidon.service.registry.ServiceRegistry.get(ServiceRegistry.java:48)
	at io.helidon.service.registry.Services.get(Services.java:156)
	at io.helidon.common.mapper.GlobalManager.mapperManager(GlobalManager.java:55)
	at io.helidon.common.mapper.MapperManager.global(MapperManager.java:56)
	at io.helidon.http.HeaderValueBase.get(HeaderValueBase.java:61)
	at io.helidon.webserver.http1.Http1Connection.route(Http1Connection.java:510)
	at io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:224)
	... 4 more
```

## Steps to reproduce

This problem can be reproduced with the following code.
```java
@Test
void test() throws Exception {

    // shares a service registry globally
    var manager = ServiceRegistryManager.start();
    manager.shutdown();

    var header = HeaderValues.create(HeaderNames.CONTENT_LENGTH, 123);

    // causes ScopeNotActiveException
    header.get(long.class);
}
```

--- a/common/context/context/src/main/java/io/helidon/common/context/Context.java
+++ b/common/context/context/src/main/java/io/helidon/common/context/Context.java
@@ -90,6 +90,16 @@ public interface Context {
     <T> void register(T instance);
 
     /**
+     * Remove an instance registration from the context.
+     *
+     * @param instance an instance to remove from context
+     * @param <T> type of the instance
+     */
+    default <T> void unregister(T instance) {
+        // do nothing by default for backward compatibility
+    }
+
+    /**
      * Register a new instance using a provided supplier. The supplier is guaranteed to be called at most once when it's
      * requested by the {@link #get(Class)} method. The returned value is then registered and the supplier is never used again.
      *
@@ -124,6 +134,17 @@ public interface Context {
      * @throws NullPointerException if {@code classifier} or registered object is {@code null}
      */
     <T> void register(Object classifier, T instance);
+
+    /**
+     * Remove an instance registration from the context.
+     *
+     * @param classifier an additional registered instance classifier
+     * @param instance an instance to remove from context
+     * @param <T> type of the instance
+     */
+    default <T> void unregister(Object classifier, T instance) {
+        // do nothing by default for backward compatibility
+    }
 
     /**
      * Registers a new instance using a provided supplier. The supplier is guarantied to be called at most once when it's

--- a/common/context/context/src/main/java/io/helidon/common/context/ListContext.java
+++ b/common/context/context/src/main/java/io/helidon/common/context/ListContext.java
@@ -56,6 +56,11 @@ class ListContext implements Context {
     }
 
     @Override
+    public <T> void unregister(T instance) {
+        registry.unregister(instance);
+    }
+
+    @Override
     public <T> void supply(Class<T> type, Supplier<T> supplier) {
         registry.supply(type, supplier);
     }
@@ -79,6 +84,13 @@ class ListContext implements Context {
         Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
         ClassifiedRegistry cr = classifiers.computeIfAbsent(classifier, k -> new ClassifiedRegistry());
         cr.register(instance);
+    }
+
+    @Override
+    public <T> void unregister(Object classifier, T instance) {
+        Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
+        ClassifiedRegistry cr = classifiers.computeIfAbsent(classifier, k -> new ClassifiedRegistry());
+        cr.unregister(instance);
     }
 
     @Override
@@ -124,7 +136,7 @@ class ListContext implements Context {
 
         // we actually want to do an instance equality
         @SuppressWarnings("ObjectEquality")
-        private void registerItem(RegisteredItem<?> item) {
+        private void registerItem(RegisteredItem<?> item, boolean register) {
             Lock l = lock.writeLock();
             try {
                 Class<?> c = item.getType();
@@ -136,7 +148,9 @@ class ListContext implements Context {
                         break;
                     }
                 }
-                content.add(item);
+                if (register) {
+                    content.add(item);
+                }
             } finally {
                 l.unlock();
             }
@@ -144,13 +158,18 @@ class ListContext implements Context {
 
         <T> void register(T instance) {
             Objects.requireNonNull(instance, "Parameter 'instance' is null!");
-            registerItem(new RegisteredInstance<>(instance));
+            registerItem(new RegisteredInstance<>(instance), true);
+        }
+
+        void unregister(Object instance) {
+            Objects.requireNonNull(instance, "Parameter 'instance' is null!");
+            registerItem(new RegisteredInstance<>(instance), false);
         }
 
         <T> void supply(Class<T> type, Supplier<T> supplier) {
             Objects.requireNonNull(type, "Parameter 'type' is null!");
             Objects.requireNonNull(supplier, "Parameter 'supplier' is null!");
-            registerItem(new RegisteredSupplier<>(type, supplier));
+            registerItem(new RegisteredSupplier<>(type, supplier), true);
         }
 
         <T> T get(Class<T> type) {

--- a/common/context/context/src/test/java/io/helidon/common/context/ContextsTest.java
+++ b/common/context/context/src/test/java/io/helidon/common/context/ContextsTest.java
@@ -33,6 +33,7 @@ import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.sameInstance;
 
 /**
  * Unit test for {@link Contexts}.
@@ -225,5 +226,27 @@ class ContextsTest {
         Contexts.clear();
 
         assertThat(Contexts.context(), is(Optional.empty()));
+    }
+
+    @Test
+    void testUnregister() {
+        Context topLevel = Contexts.globalContext();
+        Context mine = Context.builder()
+                .parent(topLevel)
+                .id("unit-test-testUnregister")
+                .build();
+
+        MyType registered = new MyType();
+        mine.register(registered);
+        MyType actual = mine.get(MyType.class).get();
+
+        assertThat(actual, sameInstance(registered));
+
+        mine.unregister(registered);
+
+        assertThat(mine.get(MyType.class), is(Optional.empty()));
+    }
+
+    private static class MyType {
     }
 }

--- a/service/registry/src/main/java/io/helidon/service/registry/GlobalServiceRegistry.java
+++ b/service/registry/src/main/java/io/helidon/service/registry/GlobalServiceRegistry.java
@@ -28,7 +28,8 @@ import io.helidon.common.context.Contexts;
  * Application wide service registry backed by {@link io.helidon.common.context.Context}.
  */
 public final class GlobalServiceRegistry {
-
+    private static final String GLOBAL_CONTEXT_CLASSIFIER = "helidon-registry-static-context";
+    private static final String GLOBAL_REGISTRY_CLASSIFIER = "helidon-registry";
     private static final ReadWriteLock RW_LOCK = new ReentrantReadWriteLock();
 
     private GlobalServiceRegistry() {
@@ -110,11 +111,20 @@ public final class GlobalServiceRegistry {
     public static ServiceRegistry registry(ServiceRegistry newGlobalRegistry) {
         RW_LOCK.writeLock().lock();
         try {
-            context().register("helidon-registry", newGlobalRegistry);
+            context().register(GLOBAL_REGISTRY_CLASSIFIER, newGlobalRegistry);
         } finally {
             RW_LOCK.writeLock().unlock();
         }
         return newGlobalRegistry;
+    }
+
+    static void unset(ServiceRegistry registry) {
+        RW_LOCK.writeLock().lock();
+        try {
+            context().unregister(GLOBAL_REGISTRY_CLASSIFIER, registry);
+        } finally {
+            RW_LOCK.writeLock().unlock();
+        }
     }
 
     private static Context context() {
@@ -123,14 +133,14 @@ public final class GlobalServiceRegistry {
         // this is the context we expect to get (and set global instances)
         return Contexts.context()
                 .orElse(globalContext)
-                .get("helidon-registry-static-context", Context.class)
+                .get(GLOBAL_CONTEXT_CLASSIFIER, Context.class)
                 .orElse(globalContext);
     }
 
     private static Optional<ServiceRegistry> current() {
         RW_LOCK.readLock().lock();
         try {
-            return context().get("helidon-registry", ServiceRegistry.class);
+            return context().get(GLOBAL_REGISTRY_CLASSIFIER, ServiceRegistry.class);
         } finally {
             RW_LOCK.readLock().unlock();
         }

--- a/service/registry/src/main/java/io/helidon/service/registry/ServiceRegistryManager.java
+++ b/service/registry/src/main/java/io/helidon/service/registry/ServiceRegistryManager.java
@@ -221,6 +221,11 @@ public final class ServiceRegistryManager {
                 return;
             }
 
+            ServiceRegistry global = GlobalServiceRegistry.registry();
+            if (global == registry) {
+                // this is the same instance, if we shut it down, global would stop working
+                GlobalServiceRegistry.unset(registry);
+            }
             registry.shutdown();
             registry = null;
         } finally {

